doc/decisions/DR016-One-to-many-result-paging.md

```diff
@@ -0,0 +1,522 @@
+    successCb: SuccessCallback,
+    errorCb,
+    limit: -1,
+    pageTokens: List[AbstractPageToken]
+): -> List[AbstractPageToken]: ...
+```
+
+E.g. Cursor object
+
+```python
+class AbstractPage:
+    def get(self) -> List[EntityReferences]:
+        ...
+    def next(self, pageSize):
+        ...
+
+# Callback taking index of batch element and cursor object.
+SuccessCallback = Callable[[int, AbstractPage], None]
+
+def relatedEntities(
+    batch: List[EntityReference],
+    relationship,
+    successCb: SuccessCallback,
+    errorCb): ...
+```
+
+#### Pros
+
+- Manager is under no obligation to consider errors in other streams
+  when implementing query behaviour, as pagination method need not be
+  synchronized.
+
+#### Pros (cursor object specific)
+
+- Individual cursors can be incremented independently, potentially
+  reducing wasted queries.
+
+#### Cons
+
+- Batching is very difficult to optimize for the manager, especially
+  after the first page.
+- Batch-optimized queries have unpredictable performance profile for
+  the host, given the wide range of potential implementations.
+
+### Outcome
+
+Given the reported difficulty of crafting queries that are both batch
+and paging-aware, we opt to go with Option 1 and split the paging API(s)
+from the batched API(s).
+
+This then leaves only two remaining solutions: (non-batched) page token
+or cursor objects.
+
+## Considering manager implementations
+
+### Using Python's DB-API standard
+
+#### Batch signature
+```python
+# Callback taking index of batch element and all results for that element.
+SuccessCallback = Callable[[int, List[EntityReference]], None]
+
+def relatedEntities(
+        batch: List[EntityReference], relationship,
+        successCb: SuccessCallback, errorCb):
+
+    cursor = get_my_connection().cursor()
+
+    # Params for query.
+    my_relationship_type: str = get_my_relationship_type(relationship)
+    entity_ref_strs = (str(ref) for ref in batch)
+
+    # Run the query.
+    placeholders = ", ".join(["?"] * len(batch))  # "?, ?, ?, ..."
+    cursor.execute(
+        f"""
+        SELECT FromRef, ToRef FROM Relationship
+        WHERE RelationshipType = ?
+        AND FromRef IN ({placeholders})
+        ORDER BY FromRef
+        """, my_relationship_type, *entity_ref_strs)
+
+    # Get the raw results
+    result_set = cursor.fetchall()
+
+    # Gather the results for each input ref.
+    result_dict = collections.defaultdict(list)
+    for result in result_set:
+        from_ref = EntityReference(result[0])
+        to_ref = EntityReference(result[1])
+        result_dict[from_ref].append(to_ref)
+
+    # Call callbacks.
+    for idx, from_ref in enumerate(batch):
+        successCb(idx, result_dict[from_ref])
+```
+
+#### Page token signature
+
+```python
+class PageToken(AbstractPageToken):
+    def __init__(self, cursor):
+        self.cursor = cursor
+
+
+def relatedEntities(
+    input: EntityReference, relationship, limit=-1, pageToken=None
```

Comment:
[consider]
```suggestion
    input: EntityReference, relationship, limit=-1, prevPageToken=None
```

Reply:
:thinking: Would it be `prev` or is it more like `current` or even `next`?  Since the token represents the page you haven't got yet.  Depends how you think about it.  Perhaps best to leave it generic?

Reply:
Yeah- maybe thats the Q - what does the token represent!